man/kvs.1
.\" Text automatically generated by txt2man
.TH untitled  "04 September 2023" "" ""
.SH NAME
\fBkvs \fP- No short description
.SH SYNOPSIS
.nf
.fam C
\fBkvs\fP <command> [<args>]
.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
No long description
.SH COMMANDS
.TP
.B
\fBkvs-enum\fP(1)
Enumerate Logical Block Namespaces on the system
.TP
.B
\fBkvs-idfy-ns\fP(1)
KV Command Set specific identify-controller
.TP
.B
\fBkvs-retrieve\fP(1)
Execute a KV Retrieve Command
.TP
.B
\fBkvs-store\fP(1)
Execute a KV Store Command
.TP
.B
\fBkvs-delete\fP(1)
Execute a KV Delete Command
.TP
.B
\fBkvs-exist\fP(1)
Execute a KV Exist Command
.TP
.B
\fBkvs-list\fP(1)
Execute a KV List Command
.RE
.PP

.SH OPTIONS
\fB--help\fP
Print the synopsis and exit
.SH EXAMPLES
Read the man page for each <command> or consult the command-line \fB--help\fP:
.PP
.nf
.fam C
    $ kvs <command> --help

.fam T
.fi
.SH SEE ALSO
Full documentation at: <https://xnvme.io/>
.SH AUTHOR
Written by Simon A. F. Lund <simon.lund@samsung.com> on behalf of Samsung


CHANGELOG.rst
.. SPDX-FileCopyrightText: Samsung Electronics Co., Ltd
..
.. SPDX-License-Identifier: BSD-3-Clause

Changelog
=========

The repository is tagged using semantic versioning, e.g. ``v1.2.3``.

The ``next`` branch contains a preview of the upcoming release. The history on
``next`` is subject to change. Upon release, ``next`` is merged with ``main``
and  and tagged ``vX.Y.Z``.

Changes are described in this file in a section matching the version tag.

Known Issues
------------

See the file named ``ISSUES`` in the root of the repository.

v0.7.1
------

* API
  - Introduced memory-mapping API (libxnvme_mem.h), that is, when the user
    manages memory and thus does not want to use xnvme_buf_alloc() as doing so
    would introduce bounce-buffers.

* CLI
  - The passthru cli-interfaces now use --cdwXY arguments instead of
    "command-files", this alignes with nvme-cli and is arguably simpler

* Backends: Linux
  - Performance improvments in io_uring and io_uring_cmd backends due to the
    addition batching via delayed submission
  - Performance improvements to libaio due to the addition of the "ring-hack"

* Backends: Windows
  - Support added for writing via "ioring" on Windows, previously only batched
    reads were possible via this interface.

* Infrastructure; github-actions, docker-images, third-party etc. bump the
  reference toolchain platforms, that is
  - Debian is now Trixie, Bookworm and Bullseye.
  - Ubuntu is now Lunar, Jammy and Focal
  - Added Rocky Linux 9.2
  - Added Oracle Linux 9.
  - Dropped CentOS 7
  - Dropped CentOS Stream 8
  - Dropped openSUSE Leap 15.3 and 15.4, Tumbleweed remains

* Removal of a bunch of deprecated files and code. General *cleanup*.

v0.7.0
------

This release got a whole lot bigger than it ought to. Aiming for more
frequent releases in the future. A bunch of changes to API,
implementation, and even the license.

Regarding the license. For compatibility with GPL-licensed software.
That is, consumption of xNVMe from GPL-licences software, then xNVMe is
changed from APACHE to BSD-3-Clause. To better determine the license,
the project has aligned with the golden practice of REUSE-compliance
(https://reuse.software/).

* API
  * Refactored to provide a single "entry-point" header
  * Headers no longer include other headers (except for the entry-point)
  * Renamed the command-line api from xnvmec to xnvme_cli

* API / Command-Sets
  * Support for FDP and DSM
  * Support for Key-Value SSDs

* Backends
  * The POSIX backend is replaced by CBI -- Common Backend Implementation
  * Instead of "POSIX" then a handful of commonly useful implementations
    are provided, being "strictly" POSIX was an impractical restriction,
    such as lacking support for features which are provided on modern
    operating systems but not defined by POSIX.

* Backends: FreeBSD
  * Support for kqueue based aio

* Backends: SPDK
  * Now has options to control Admin and I/O command timeout
  * Removed patches no longer needed (available upstream)

* Backends: vfio
  * Removed as subproject and instead links "properly" with the
    system-provided libvfn

* tools
  * switched fio from "external" to "internal" / "upstream" fio engine

For additional details then please have a look at the commit-history,
the v0.7.0 related issue on GitHUB and PR.

v0.6.0
------

A handful of improvements for Windows, additional steps toward removing
third-party vendoring / bundling, addition of experimental tunable knobs for
io_uring/SQPOLL, and several improvements to testing and CI infra.

* API
  - Removed SLIST from API, although "sys/queue.h" are commonly available on
    Linux/FreeBSD, then they are not part of toolchain on Windows.

* Third-party
  - Bumped SPDK to v22.09, and with that removed mutliple out-of-tree patches
    for DPDK.

* CLI
  - The xNVMe command-line library (libxnvmec) and all the cli-tools using it
    are refactored to use common sets of command-line arguments. Along with
    this came a consistent set of CLI-arguments for admin/sync/async.

* Backends
  - ramdisk: The ramdisk got support for write-zeroes, iovec payloads and added
    to CI testing.
  - linux: support for a buffer-allocator using HUGEPAGES and tunable knobs for
    controlling the behavior of io_uring SQPOLL via environment variables.
  - windows: support for the experimental IORING Windows SPDK API and support
    for block devices (SCSI and SATA).
  - spdk: when controllers are re-used for device-handles, events are
    processed as a means to check whether the controller is still "alive"

* CI
  - scan-build now runs for each PR
  - basic tests are now running post-building testing using the RAMDISK
  - Testing of fabrics with TCP transport is now part of the setup

v0.5.0
------

A bit of expansion in the application of xNVMe with support for macOS, a
ramdisk backend, revival of Python language bindings and a refresh of the docs
on NVMe-over-Fabrics.

* API
  - Removed helpers for SGL
  - Add 'subnqn' to 'xnvme_ident', useful for NVMe-oF
  - Add 'hostnqn' to 'opts', useful for NVMe-oF
  - Add support directive-receive and write-with-directives
  - Spec adjusted for NVMe 2.0, still more work needed in this area

* Third-party
  - Bumped fio to v3.32
  - Bumped SPDK to v22.05
  - Removed liburing, now relies on on-system library instead of
    vendoring/bundling, documentation is updated to assist with library
    installation

* Tooling
  - Re-working testing using cijoe 0.9+, that is, switching to CIJOE/pytest for
    testing an CIJOE workflows for instrumentation
  - liburing is no longer bundled with xNVMe, that is, xNVMe now links with
    liburing as discovered on the system via pkg-config. This is done to avoid
    symbol collisions for applications linking or loading liburing and xNVMe.

* be:linux:async:libaio
  - When 'opts.poll_io' is set then poke() will return immmediatly and now wait
    for completions. This allows the use of trading CPU for more IOPS and lower

tools/kvs.c
// SPDX-FileCopyrightText: Samsung Electronics Co., Ltd
//
// SPDX-License-Identifier: BSD-3-Clause

#include <errno.h>
#include <libxnvme.h>

int
enumerate_cb(struct xnvme_dev *dev, void *cb_args)
{
	uint32_t *ns_count_ref = cb_args;
	const struct xnvme_ident *ident;

	ident = xnvme_dev_get_ident(dev);
	if (ident->csi != XNVME_SPEC_CSI_KV) {
		return XNVME_ENUMERATE_DEV_CLOSE;
	}

	if (*ns_count_ref == 0) {
		fprintf(stdout, "\n");
	}

	fprintf(stdout, "  - {");
	xnvme_ident_yaml(stdout, ident, 0, ", ", 0);
	fprintf(stdout, "}\n");

	*ns_count_ref = *ns_count_ref + 1;

	return XNVME_ENUMERATE_DEV_CLOSE;
}

static int
cmd_enumerate(struct xnvme_cli *cli)
{
	struct xnvme_opts opts = {0};
	uint32_t ns_count = 0;
	int err = 0;

	err = xnvme_cli_to_opts(cli, &opts);
	if (err) {
		xnvme_cli_perr("xnvme_cli_to_opts()", err);
		return err;
	}

	fprintf(stdout, "xnvme_enumeration:");

	err = xnvme_enumerate(cli->args.sys_uri, &opts, *enumerate_cb, &ns_count);
	if (err) {
		xnvme_cli_perr("xnvme_enumerate()", err);
		return err;
	}

	if (ns_count == 0) {
		fprintf(stdout, "~\n");
	}

	return 0;
}

static inline int
cmd_idfy_ns(struct xnvme_cli *cli)
{
	struct xnvme_dev *dev = cli->args.dev;
	struct xnvme_cmd_ctx ctx = xnvme_cmd_ctx_from_dev(dev);
	uint32_t nsid = xnvme_dev_get_nsid(cli->args.dev);
	enum xnvme_spec_csi csi = XNVME_SPEC_CSI_KV;
	struct xnvme_spec_kvs_idfy *idfy = NULL;
	int err;

	if (!cli->given[XNVME_CLI_OPT_NSID]) {
		nsid = xnvme_dev_get_nsid(cli->args.dev);
	}

	xnvme_cli_pinf("xnvme_adm_idfy_ns: {nsid: 0x%x, csi: %s}", nsid, xnvme_spec_csi_str(csi));

	idfy = xnvme_buf_alloc(dev, sizeof(*idfy));
	if (!idfy) {
		err = -errno;
		xnvme_cli_perr("xnvme_buf_alloc()", err);
		goto exit;
	}

	err = xnvme_adm_idfy_ns_csi(&ctx, nsid, XNVME_SPEC_CSI_KV, &idfy->base);
	if (err || xnvme_cmd_ctx_cpl_status(&ctx)) {
		xnvme_cli_perr("xnvme_adm_idfy_ns_csi()", err);
		xnvme_cmd_ctx_pr(&ctx, XNVME_PR_DEF);
		err = err ? err : -EIO;
		goto exit;
	}

	xnvme_spec_kvs_idfy_ns_pr(&idfy->ns, XNVME_PR_DEF);

	if (cli->args.data_output) {
		xnvme_cli_pinf("Dumping to: '%s'", cli->args.data_output);
		err = xnvme_buf_to_file((char *)idfy, sizeof(*idfy), cli->args.data_output);
		if (err) {
			xnvme_cli_perr("xnvme_buf_to_file()", err);
		}
	}

exit:
	xnvme_buf_free(dev, idfy);

	return err;
}

static int
cmd_retrieve(struct xnvme_cli *cli)
{
	struct xnvme_dev *dev = cli->args.dev;
	struct xnvme_cmd_ctx ctx = xnvme_cmd_ctx_from_dev(dev);
	uint8_t nsid = cli->args.nsid;
	void *dbuf = NULL;
	size_t dbuf_nbytes = 4096;
	int err;

	if (!cli->given[XNVME_CLI_OPT_NSID]) {
		nsid = xnvme_dev_get_nsid(cli->args.dev);
	}

	xnvme_cli_pinf("Alloc/clear dbuf, dbuf_nbytes: %zu", dbuf_nbytes);

	dbuf = xnvme_buf_alloc(dev, dbuf_nbytes);
	if (!dbuf) {
		err = -errno;
		xnvme_cli_perr("xnvme_buf_alloc()", err);
		goto exit;
	}
	memset(dbuf, 0, dbuf_nbytes);

	xnvme_cli_pinf("Sending xnvme_kvs_retrieve command");
	err = xnvme_kvs_retrieve(&ctx, nsid, cli->args.kv_key, strlen(cli->args.kv_key), dbuf,
				 dbuf_nbytes, 0);
	if (err || xnvme_cmd_ctx_cpl_status(&ctx)) {
		xnvme_cli_perr("xnvme_kvs_retrieve()", err);
		xnvme_cmd_ctx_pr(&ctx, XNVME_PR_DEF);
		err = err ? err : -EIO;
		goto exit;
	}

	if (cli->args.data_output) {
		xnvme_cli_pinf("dumping to: '%s'", cli->args.data_output);
		err = xnvme_buf_to_file(dbuf, dbuf_nbytes, cli->args.data_output);
		if (err) {
			xnvme_cli_perr("xnvme_buf_to_file()", err);
		}
	} else {
		xnvme_cli_pinf("KV Value retrieved: '%s'", dbuf);
	}

exit:
	xnvme_buf_free(dev, dbuf);

	return err;
}

static int
cmd_store(struct xnvme_cli *cli)
{
	struct xnvme_dev *dev = cli->args.dev;
	struct xnvme_cmd_ctx ctx = xnvme_cmd_ctx_from_dev(dev);
	uint8_t nsid = cli->args.nsid;
	uint8_t opt = 0x0;
	void *dbuf = NULL;
	size_t dbuf_nbytes = strlen(cli->args.kv_val) + 1;
	int err;

	if (!cli->given[XNVME_CLI_OPT_NSID]) {
		nsid = xnvme_dev_get_nsid(cli->args.dev);
	}
	xnvme_cli_pinf("Alloc/fill dbuf, dbuf_nbytes: %zu", dbuf_nbytes);

	if (cli->args.kv_store_add) {
		opt = opt | XNVME_KVS_STORE_OPT_DONT_STORE_IF_KEY_EXISTS;
	} else if (cli->args.kv_store_update) {
		opt = opt | XNVME_KVS_STORE_OPT_DONT_STORE_IF_KEY_NOT_EXISTS;
	}

	dbuf = xnvme_buf_alloc(dev, dbuf_nbytes);
	if (!dbuf) {
		err = -errno;
		xnvme_cli_perr("xnvme_buf_alloc()", err);
		goto exit;
	}

	memcpy(dbuf, cli->args.kv_val, dbuf_nbytes);

	err = xnvme_kvs_store(&ctx, nsid, cli->args.kv_key, strlen(cli->args.kv_key), dbuf,
			      dbuf_nbytes, opt);
	if (err || xnvme_cmd_ctx_cpl_status(&ctx)) {
		xnvme_cli_perr("xnvme_kvs_store()", err);
		xnvme_cmd_ctx_pr(&ctx, XNVME_PR_DEF);
		err = err ? err : -EIO;
		goto exit;
	}

	xnvme_cli_pinf("KV Value is written under a key: \"%s\"\n", cli->args.kv_key);
exit:
	xnvme_buf_free(dev, dbuf);

	return err;
}

static int
cmd_delete(struct xnvme_cli *cli)
{
	struct xnvme_dev *dev = cli->args.dev;
	struct xnvme_cmd_ctx ctx = xnvme_cmd_ctx_from_dev(dev);
	uint8_t nsid = cli->args.nsid;
	int err;

	if (!cli->given[XNVME_CLI_OPT_NSID]) {
		nsid = xnvme_dev_get_nsid(cli->args.dev);
	}

	xnvme_cli_pinf("Sending xnvme_kvs_delete command");

	err = xnvme_kvs_delete(&ctx, nsid, cli->args.kv_key, strlen(cli->args.kv_key));
	if (err || xnvme_cmd_ctx_cpl_status(&ctx)) {
		xnvme_cli_perr("xnvme_kv_delete()", err);
		xnvme_cmd_ctx_pr(&ctx, XNVME_PR_DEF);
		err = err ? err : -EIO;
		goto exit;
	}

	xnvme_cli_pinf("Removed key: '%s'", cli->args.kv_key);

exit:
	return err;
}

static int
cmd_exist(struct xnvme_cli *cli)
{
	struct xnvme_dev *dev = cli->args.dev;
	struct xnvme_cmd_ctx ctx = xnvme_cmd_ctx_from_dev(dev);
	uint8_t nsid = cli->args.nsid;
	int err;

	if (!cli->given[XNVME_CLI_OPT_NSID]) {
		nsid = xnvme_dev_get_nsid(cli->args.dev);
	}

	xnvme_cli_pinf("Sending xnvme_kvs_exist command");
	err = xnvme_kvs_exist(&ctx, nsid, cli->args.kv_key, strlen(cli->args.kv_key));
	if (err || xnvme_cmd_ctx_cpl_status(&ctx)) {
		xnvme_cli_perr("xnvme_kvs_exist()", err);
		xnvme_cmd_ctx_pr(&ctx, XNVME_PR_DEF);
		err = err ? err : -EIO;
		goto exit;
	}

	xnvme_cli_pinf("Key '%s' exists", cli->args.kv_key);

exit:
	return err;
}

static int
cmd_list(struct xnvme_cli *cli)
{
	struct xnvme_dev *dev = cli->args.dev;
	struct xnvme_cmd_ctx ctx = xnvme_cmd_ctx_from_dev(dev);
	uint8_t nsid = cli->args.nsid;
	void *dbuf = NULL;
	size_t stdio_wrtn, dbuf_nbytes = 4096;
	int err;

	if (!cli->given[XNVME_CLI_OPT_NSID]) {
		nsid = xnvme_dev_get_nsid(cli->args.dev);
	}

	XNVME_DEBUG("Alloc/clear dbuf, dbuf_nbytes: %zu", dbuf_nbytes);

	dbuf = xnvme_buf_alloc(dev, dbuf_nbytes);
	if (!dbuf) {
		err = -errno;
		xnvme_cli_perr("xnvme_buf_alloc()", err);
		goto exit;
	}
	memset(dbuf, 0, dbuf_nbytes);

	XNVME_DEBUG("Sending xnvme_kvs_list command");

	err = xnvme_kvs_list(&ctx, nsid, cli->args.kv_key, strlen(cli->args.kv_key), dbuf,
			     dbuf_nbytes);
	if (err || xnvme_cmd_ctx_cpl_status(&ctx)) {
		xnvme_cli_perr("xnvme_kvs_list()", err);
		xnvme_cmd_ctx_pr(&ctx, XNVME_PR_DEF);
		err = err ? err : -EIO;
		goto exit;
	}

	XNVME_DEBUG("KV List retrieved with %u keys\n", *(uint32_t *)dbuf);

	stdio_wrtn = fwrite(dbuf, sizeof(uint8_t), dbuf_nbytes, stdout);

	if (stdio_wrtn != dbuf_nbytes) {
		err = (int)stdio_wrtn;
		xnvme_cli_perr("fwrite()", err);
		err = err ? err : -EIO;
		goto exit;
	}

	return 0;
exit:
	xnvme_buf_free(dev, dbuf);

	return err;
}

//
// Command-Line Interface (CLI) definition
//

static struct xnvme_cli_sub g_subs[] = {
	{"enum",
	 "Enumerate Logical Block Namespaces on the system",
	 "Enumerate Logical Block Namespaces on the system",
	 cmd_enumerate,
	 {
		 {XNVME_CLI_OPT_NON_POSA_TITLE, XNVME_CLI_SKIP},

		 {XNVME_CLI_OPT_SYS_URI, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_FLAGS, XNVME_CLI_LOPT},

		 {XNVME_CLI_OPT_DEV_NSID, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_BE, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_ADMIN, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_SYNC, XNVME_CLI_LOPT},
	 }},
	{"idfy-ns",
	 "Zoned Command Set specific identify-controller",
	 "KV Command Set specific identify-controller",
	 cmd_idfy_ns,
	 {
		 {XNVME_CLI_OPT_NON_POSA_TITLE, XNVME_CLI_SKIP},

		 {XNVME_CLI_OPT_URI, XNVME_CLI_POSA},
		 {XNVME_CLI_OPT_NSID, XNVME_CLI_LOPT},

		 {XNVME_CLI_OPT_DEV_NSID, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_BE, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_ADMIN, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_SYNC, XNVME_CLI_LOPT},
	 }},
	{"retrieve",
	 "Execute a KV Retrieve Command",
	 "Execute a KV Retrieve Command",
	 cmd_retrieve,
	 {
		 {XNVME_CLI_OPT_NON_POSA_TITLE, XNVME_CLI_SKIP},

		 {XNVME_CLI_OPT_URI, XNVME_CLI_POSA},
		 {XNVME_CLI_OPT_NSID, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_KV_KEY, XNVME_CLI_LREQ},
		 {XNVME_CLI_OPT_DATA_OUTPUT, XNVME_CLI_LOPT},

		 {XNVME_CLI_OPT_DEV_NSID, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_BE, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_ADMIN, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_SYNC, XNVME_CLI_LOPT},
	 }},
	{"store",
	 "Execute a KV Store Command",
	 "Execute a KV Store Command",
	 cmd_store,
	 {
		 {XNVME_CLI_OPT_NON_POSA_TITLE, XNVME_CLI_SKIP},

		 {XNVME_CLI_OPT_URI, XNVME_CLI_POSA},
		 {XNVME_CLI_OPT_NSID, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_KV_KEY, XNVME_CLI_LREQ},
		 {XNVME_CLI_OPT_KV_VAL, XNVME_CLI_LREQ},

		 {XNVME_CLI_OPT_KV_STORE_UPDATE, XNVME_CLI_LFLG},
		 {XNVME_CLI_OPT_KV_STORE_ADD, XNVME_CLI_LFLG},
		 {XNVME_CLI_OPT_KV_STORE_COMPRESS, XNVME_CLI_LFLG},

		 {XNVME_CLI_OPT_DEV_NSID, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_BE, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_ADMIN, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_SYNC, XNVME_CLI_LOPT},
	 }},
	{"delete",
	 "Execute a KV Delete Command",
	 "Execute a KV Delete Command",
	 cmd_delete,
	 {
		 {XNVME_CLI_OPT_NON_POSA_TITLE, XNVME_CLI_SKIP},

		 {XNVME_CLI_OPT_URI, XNVME_CLI_POSA},
		 {XNVME_CLI_OPT_NSID, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_KV_KEY, XNVME_CLI_LREQ},

		 {XNVME_CLI_OPT_DEV_NSID, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_BE, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_ADMIN, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_SYNC, XNVME_CLI_LOPT},
	 }},
	{"exist",
	 "Execute a KV Exist Command",
	 "Execute a KV Exist Command",
	 cmd_exist,
	 {
		 {XNVME_CLI_OPT_NON_POSA_TITLE, XNVME_CLI_SKIP},

		 {XNVME_CLI_OPT_URI, XNVME_CLI_POSA},
		 {XNVME_CLI_OPT_NSID, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_KV_KEY, XNVME_CLI_LREQ},

		 {XNVME_CLI_OPT_DEV_NSID, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_BE, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_ADMIN, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_SYNC, XNVME_CLI_LOPT},
	 }},
	{"list",
	 "Execute a KV List Command",
	 "Execute a KV List Command",
	 cmd_list,
	 {
		 {XNVME_CLI_OPT_NON_POSA_TITLE, XNVME_CLI_SKIP},

		 {XNVME_CLI_OPT_URI, XNVME_CLI_POSA},
		 {XNVME_CLI_OPT_NSID, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_KV_KEY, XNVME_CLI_LREQ},

		 {XNVME_CLI_OPT_DEV_NSID, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_BE, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_ADMIN, XNVME_CLI_LOPT},
		 {XNVME_CLI_OPT_SYNC, XNVME_CLI_LOPT},
	 }},
};

static struct xnvme_cli g_cli = {
	.title = "KV Utility",
	.descr_short = "Retrieve KV value",
	.subs = g_subs,
	.nsubs = sizeof g_subs / sizeof(*g_subs),
};

int
main(int argc, char **argv)
{
	return xnvme_cli_run(&g_cli, argc, argv, XNVME_CLI_INIT_DEV_OPEN);
}
